File: packages/sdk-python/beam_directory/frames.py
"""Intent and Result frame creation and validation for the Beam Protocol."""

from __future__ import annotations

import json
from datetime import datetime, timezone
from typing import Any, Optional

from .identity import BeamIdentity
from .types import BeamIdString, IntentFrame, ResultFrame

MAX_FRAME_SIZE = 1024        # bytes
REPLAY_WINDOW_MS = 300_000  # 5 minutes


def _canonical_json(data: dict[str, Any]) -> str:
    """Produce deterministic JSON (sorted keys, no spaces)."""
    return json.dumps(data, sort_keys=True, separators=(",", ":"))


def _intent_signature_payload(
    *,
    from_id: BeamIdString,
    to_id: BeamIdString,
    intent: str,
    payload: dict[str, Any],
    timestamp: str,
    nonce: str,
) -> str:
    """Match the TypeScript SDK's on-wire signature payload exactly."""
    return json.dumps(
        {
            "type": "intent",
            "from": from_id,
            "to": to_id,
            "intent": intent,
            "payload": payload,
            "timestamp": timestamp,
            "nonce": nonce,
        },
        separators=(",", ":"),
    )


def create_intent_frame(
    intent: str,
    from_id: BeamIdString,
    to_id: BeamIdString,
    params: Optional[dict[str, Any]] = None,
    identity: Optional[BeamIdentity] = None,
) -> IntentFrame:
    """
    Create a signed IntentFrame.

    If *identity* is provided, the frame is signed. Signature covers a
    canonical JSON serialisation of all fields except *signature* itself.
    """
    frame = IntentFrame(
        v="1",
        intent=intent,
        from_id=from_id,
        to_id=to_id,
        params=params or {},
        nonce=BeamIdentity.generate_nonce(),
        timestamp=datetime.now(timezone.utc).isoformat().replace("+00:00", "Z"),
    )

    if identity is not None:
        frame.signature = identity.sign(
            _intent_signature_payload(
                from_id=frame.from_id,
                to_id=frame.to_id,
                intent=frame.intent,
                payload=frame.params,
                timestamp=frame.timestamp,
                nonce=frame.nonce,
            )
        )

    return frame


def create_result_frame(
    success: bool,
    nonce: str,
    payload: Optional[dict[str, Any]] = None,
    error: Optional[str] = None,
    error_code: Optional[str] = None,
    latency: Optional[int] = None,
    identity: Optional[BeamIdentity] = None,
) -> ResultFrame:
    """Create a ResultFrame (optionally signed)."""
    frame = ResultFrame(
        v="1",
        success=success,
        nonce=nonce,
        timestamp=datetime.now(timezone.utc).isoformat().replace("+00:00", "Z"),
        payload=payload,
        error=error,
        error_code=error_code,
        latency=latency,
    )

    if identity is not None:
        canonical = _canonical_json(frame.to_dict())
        frame.signature = identity.sign(canonical)

    return frame


def validate_intent_frame(
    frame: IntentFrame,
    public_key_base64: Optional[str] = None,
) -> list[str]:
    """
    Validate an IntentFrame. Returns a list of error strings (empty = valid).

    If *public_key_base64* is given, also verifies the signature.
    """
    errors: list[str] = []

    if frame.v != "1":
        errors.append(f"Unknown frame version: {frame.v!r}")

    if not frame.intent:
        errors.append("intent is required")

    if not frame.from_id:
        errors.append("from_id is required")
    elif BeamIdentity.parse_beam_id(frame.from_id) is None:
        errors.append(f"Invalid from_id: {frame.from_id!r}")

    if not frame.to_id:
        errors.append("to_id is required")
    elif BeamIdentity.parse_beam_id(frame.to_id) is None:
        errors.append(f"Invalid to_id: {frame.to_id!r}")

    if not frame.nonce:
        errors.append("nonce is required")

    if not frame.timestamp:
        errors.append("timestamp is required")
    else:
        try:
            ts = datetime.fromisoformat(frame.timestamp.replace("Z", "+00:00"))
            now = datetime.now(timezone.utc)
            age_ms = abs((now - ts).total_seconds() * 1000)
            if age_ms > REPLAY_WINDOW_MS:
                errors.append(f"Frame timestamp out of replay window ({age_ms:.0f}ms)")
        except ValueError:
            errors.append(f"Invalid timestamp: {frame.timestamp!r}")

    # Size check — use canonical form
    size = len(_canonical_json(frame.to_dict()).encode("utf-8"))
    if size > MAX_FRAME_SIZE:
        errors.append(f"Frame too large: {size} bytes (max {MAX_FRAME_SIZE})")

    # Signature check
    if public_key_base64 and frame.signature:
        if not BeamIdentity.verify(
            _intent_signature_payload(
                from_id=frame.from_id,
                to_id=frame.to_id,
                intent=frame.intent,
                payload=frame.params,
                timestamp=frame.timestamp,
                nonce=frame.nonce,
            ),
            frame.signature,
            public_key_base64,
        ):
            errors.append("Signature verification failed")

    return errors


def validate_result_frame(frame: ResultFrame, public_key_base64: Optional[str] = None) -> list[str]:
    """Validate a ResultFrame. Returns list of errors (empty = valid)."""
    errors: list[str] = []

    if frame.v != "1":
        errors.append(f"Unknown frame version: {frame.v!r}")

    if not frame.nonce:
        errors.append("nonce is required")

    if not frame.timestamp:
        errors.append("timestamp is required")

    if not frame.success and not frame.error:
        errors.append("error message required when success=False")

    if public_key_base64:
        if not frame.signature:
            errors.append("signature is required when public_key_base64 is provided")
        else:
            unsigned = frame.to_dict()
            unsigned.pop("signature", None)
            if not BeamIdentity.verify(_canonical_json(unsigned), frame.signature, public_key_base64):
                errors.append("Signature verification failed")

    return errors

File: packages/sdk-typescript/src/frames.ts
import { randomUUID, createPrivateKey, sign } from 'node:crypto'
import type { IntentFrame, ResultFrame, BeamIdString } from './types.js'
import { BeamIdentity } from './identity.js'

export const MAX_FRAME_SIZE = 4 * 1024  // 4KB hard limit
export const REPLAY_WINDOW_MS = 5 * 60 * 1000  // 5 minutes

export function createIntentFrame(
  options: {
    intent: string
    from: BeamIdString
    to: BeamIdString
    payload?: Record<string, unknown>
  },
  identity: BeamIdentity
): IntentFrame {
  const frame: IntentFrame = {
    v: '1',
    intent: options.intent,
    from: options.from,
    to: options.to,
    payload: options.payload ?? {},
    nonce: randomUUID(),
    timestamp: new Date().toISOString()
  }
  return signFrame(frame, identity.export().privateKeyBase64)
}

export function signFrame(frame: IntentFrame, privateKeyBase64: string): IntentFrame {
  const signedPayload = JSON.stringify({
    type: 'intent',
    from: frame.from,
    to: frame.to,
    intent: frame.intent,
    payload: frame.payload,
    timestamp: frame.timestamp,
    nonce: frame.nonce,
  })
  const privateKey = createPrivateKey({
    key: Buffer.from(privateKeyBase64, 'base64'),
    format: 'der',
    type: 'pkcs8',
  })
  frame.signature = sign(null, Buffer.from(signedPayload, 'utf8'), privateKey).toString('base64')
  return frame
}

export function createResultFrame(
  options: {
    nonce: string
    success: boolean
    payload?: Record<string, unknown>
    error?: string
    errorCode?: string
    latency?: number
  },
  identity: BeamIdentity
): ResultFrame {
  const frame: ResultFrame = {
    v: '1',
    success: options.success,
    nonce: options.nonce,
    timestamp: new Date().toISOString(),
    ...(options.payload !== undefined && { payload: options.payload }),
    ...(options.error !== undefined && { error: options.error }),
    ...(options.errorCode !== undefined && { errorCode: options.errorCode }),
    ...(options.latency !== undefined && { latency: options.latency })
  }
  frame.signature = identity.sign(canonicalizeFrame(frame as unknown as Record<string, unknown>))
  return frame
}

export function validateIntentFrame(
  frame: unknown,
  senderPublicKey: string
): { valid: boolean; error?: string } {
  if (!frame || typeof frame !== 'object') {
    return { valid: false, error: 'Frame must be an object' }
  }
  const f = frame as Record<string, unknown>

  if (f['v'] !== '1') return { valid: false, error: 'Invalid protocol version' }
  if (typeof f['intent'] !== 'string' || !f['intent']) return { valid: false, error: 'Missing or empty intent' }
  if (typeof f['from'] !== 'string' || !BeamIdentity.parseBeamId(f['from'])) {
    return { valid: false, error: 'Invalid from Beam ID' }
  }
  if (typeof f['to'] !== 'string' || !BeamIdentity.parseBeamId(f['to'])) {
    return { valid: false, error: 'Invalid to Beam ID' }
  }
  if (typeof f['nonce'] !== 'string' || !f['nonce']) return { valid: false, error: 'Missing nonce' }
  if (typeof f['timestamp'] !== 'string') return { valid: false, error: 'Missing timestamp' }

  const payload = normalizeIntentPayload(f)
  if (!payload) {
    return { valid: false, error: 'Payload must be an object' }
  }

  const size = Buffer.byteLength(JSON.stringify(frame), 'utf8')
  if (size > MAX_FRAME_SIZE) {
    return { valid: false, error: `Frame size ${size} exceeds limit of ${MAX_FRAME_SIZE} bytes` }
  }

  const frameTime = new Date(f['timestamp'] as string).getTime()
  if (isNaN(frameTime)) return { valid: false, error: 'Invalid timestamp format' }
  if (Math.abs(Date.now() - frameTime) > REPLAY_WINDOW_MS) {
    return { valid: false, error: 'Frame timestamp outside replay window (±5 minutes)' }
  }

  if (typeof f['signature'] !== 'string') return { valid: false, error: 'Missing signature' }
  const signedPayload = JSON.stringify({
    type: 'intent',
    from: f['from'],
    to: f['to'],
    intent: f['intent'],
    payload,
    timestamp: f['timestamp'],
    nonce: f['nonce'],
  })
  if (!BeamIdentity.verify(signedPayload, f['signature'] as string, senderPublicKey)) {
    return { valid: false, error: 'Signature verification failed' }
  }

  return { valid: true }
}

function normalizeIntentPayload(frame: Record<string, unknown>): Record<string, unknown> | null {
  const payload = frame['payload']
  if (payload && typeof payload === 'object' && !Array.isArray(payload)) {
    return payload as Record<string, unknown>
  }

  const params = frame['params']
  if (params && typeof params === 'object' && !Array.isArray(params)) {
    frame['payload'] = params
    return params as Record<string, unknown>
  }

  return null
}

export function validateResultFrame(
  frame: unknown,
  senderPublicKey: string
): { valid: boolean; error?: string } {
  if (!frame || typeof frame !== 'object') {
    return { valid: false, error: 'Frame must be an object' }
  }
  const f = frame as Record<string, unknown>

  if (f['v'] !== '1') return { valid: false, error: 'Invalid protocol version' }
  if (typeof f['success'] !== 'boolean') return { valid: false, error: 'Missing success boolean' }
  if (typeof f['nonce'] !== 'string' || !f['nonce']) return { valid: false, error: 'Missing nonce' }
  if (typeof f['timestamp'] !== 'string') return { valid: false, error: 'Missing timestamp' }
  if (typeof f['signature'] !== 'string') return { valid: false, error: 'Missing signature' }

  const { signature, ...unsigned } = f
  if (!BeamIdentity.verify(canonicalizeFrame(unsigned), signature as string, senderPublicKey)) {
    return { valid: false, error: 'Signature verification failed' }
  }

  return { valid: true }
}

export function canonicalizeFrame(frame: Record<string, unknown>): string {
  return JSON.stringify(deepSortKeys(frame))
}

function deepSortKeys(value: unknown): unknown {
  if (Array.isArray(value)) return value.map(deepSortKeys)
  if (value !== null && typeof value === 'object') {
    const sorted: Record<string, unknown> = {}
    for (const key of Object.keys(value as object).sort()) {
      sorted[key] = deepSortKeys((value as Record<string, unknown>)[key])
    }
    return sorted
  }
  return value
}
